binaries/cuprated/src/config.rs

```diff
@@ -1 +1,145 @@
 //! cuprated config
+use std::{
+    fs::{read_to_string, File},
+    io,
+    path::Path,
+    time::Duration,
+};
+
+use clap::Parser;
+use serde::{Deserialize, Serialize};
+
+use cuprate_consensus::ContextConfig;
+use cuprate_helper::{fs::CUPRATE_CONFIG_DIR, network::Network};
+use cuprate_p2p::block_downloader::BlockDownloaderConfig;
+use cuprate_p2p_core::ClearNet;
+
+mod args;
+mod default;
+mod p2p;
+mod storage;
+mod tracing_config;
+
+use p2p::P2PConfig;
+use storage::StorageConfig;
+use tracing_config::TracingConfig;
+
+/// The default name of Cuprate's config file.
+const DEFAULT_CONFIG_FILE_NAME: &str = "Cuprate.toml";
+
+/// Reads the args & config file, returning a [`Config`].
+pub fn read_config_and_args() -> Config {
+    let args = args::Args::parse();
+
+    let config: Config = if let Some(config_file) = &args.config_file {
+        // If a config file was set in the args try read it and exit if we can't.
+        match Config::read_from_file(config_file) {
+            Ok(config) => config,
+            Err(e) => {
+                tracing::error!("Failed to read config from file: {}", e);
+                std::process::exit(1);
+            }
+        }
+    } else {
+        // First attempt to read the config file from the current directory.
+        std::env::current_dir()
+            .map_err(Into::into)
+            .and_then(Config::read_from_file)
+            .inspect_err(|e| tracing::debug!("Failed to read config from current dir: {e}"))
+            // otherwise try the main config directory.
+            .or_else(|_| {
+                let file = CUPRATE_CONFIG_DIR.join(DEFAULT_CONFIG_FILE_NAME);
+                Config::read_from_file(file)
+            })
+            .inspect_err(|e| {
+                tracing::debug!("Failed to read config from config dir: {e}");
+                tracing::warn!("Failed to find/read config file, using default config.");
+            })
+            .unwrap_or_default()
+    };
+
+    args.apply_args(config)
+}
+
+/// The config for all of Cuprate.
+#[derive(Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields, default)]
+pub struct Config {
+    /// The network we should run on.
+    network: Network,
+
+    /// [`tracing`] config.
+    tracing: TracingConfig,
+
+    /// The P2P network config.
+    p2p: P2PConfig,
+
+    /// The Storage config
```

Comment:
This is habit from fulfilling the lint requiring comments, I don't think it's enabled here though, I am still going to leave them here though for if that ever happens. FWIW you do this too.

Reply:
> FWIW you do this too

Fair enough, I have done this in the past (in personal/company projects and this codebase) for the exact same reason and I regret it, I try not to do it anymore.

Lots of the criticism I'll give code is because I see a pattern that I've personally done in the past but have since then realized it was a mistake.

> This is habit from fulfilling the lint requiring comments, I don't think it's enabled here though, I am still going to leave them here though for if that ever happens.

I'd still prefer them undocumented since _if_ we enable the lint (I'm less confident we should do this now for the reasons above) then it will show us the fields that are not documented. Duplicating the information is close to doing `#[allow(missing_docs)]` but worse since it adds noise to docs.

Can we compromise and agree that our codebase should not have docs/comments like this in the future?

Reply:
In case you saw it don't mind my prior message I just deleted

Reply:
> Duplicating the information is close to doing #[allow(missing_docs)] but worse since it adds noise to docs.

I think this is something that isn't substantial at all. Most developers will not pay attention to the duplication of info nor does it cost any issue in development.

>  Can we compromise and agree that our codebase should not have docs/comments like this in the future?

My stance is we should stay lenient on that. I don't think this type of review point is even appropriate for other contributors. It rely more of the semantic perfection than actual correctness improvement IMO.

Reply:
I wrote a long response but it's not worth it. This nit can be ignored/resolved.